cmd/ndm_daemonset/probe/addhandler.go

```diff
@@ -207,12 +256,344 @@ func (pe *ProbeEvent) createBlockDeviceResourceIfNoHolders(bd blockdevice.BlockD
+				return true, nil
+			} else {
+				// should never reach this case
+				klog.Error("unreachable state")
+				return false, fmt.Errorf("unreachable state")
+			}
+		}
+	}
+
+	legacyUUID, isVirt := generateLegacyUUID(bd)
+	existingLegacyBD := pe.Controller.GetExistingBlockDeviceResource(bdAPIList, legacyUUID)
+
+	// check if any blockdevice exist with the annotation, if yes, that will be used.
+	// This is to handle the case where device comes at the same path of an earlier device
+	if r := getExistingBDWithFsUuid(bd, bdAPIList); r != nil {
+		existingLegacyBD = r
+	}
+
+	// if existingBD is nil. i.e no blockdevice exist with the uuid or fsuuid annotation, then we create
+	// the resource.
+	if existingLegacyBD == nil {
+		// create device with fs annotation and legacy annotation
+		// the custom create / update method should be called here
+		// no further processing is required
+		bd.UUID = legacyUUID
+		err := pe.createOrUpdateWithFSUUID(bd, existingLegacyBD)
+		return false, err
+	}
+
+	if existingLegacyBD.Status.ClaimState != apis.BlockDeviceUnclaimed {
+		// update resource with legacy and fsuuid annotation
+		// further processing is not required
+		bd.UUID = existingLegacyBD.Name
+		err := pe.createOrUpdateWithFSUUID(bd, existingLegacyBD)
+		return false, err
+	}
+
+	if isVirt {
+		// update the resource with fs and legacy annotation
+		bd.UUID = existingLegacyBD.Name
+		err := pe.createOrUpdateWithFSUUID(bd, existingLegacyBD)
+		return false, err
+	} else {
+		// should never reach this case.
+		klog.Error("unreachable state")
+		return false, fmt.Errorf("unreachable state")
+	}
+}
+
+// isParentDeviceInUse checks if the parent device of a given device is in use.
+// The check is made only if the device is a partition
+func (pe *ProbeEvent) isParentDeviceInUse(bd blockdevice.BlockDevice) (bool, error) {
+	if bd.DeviceAttributes.DeviceType != blockdevice.BlockDeviceTypePartition {
+		return false, nil
+	}
+
+	parentBD, ok := pe.Controller.BDHierarchy[bd.DependentDevices.Parent]
+	if !ok {
+		return false, fmt.Errorf("cannot find parent device of %s", bd.DevPath)
+	}
+
+	return parentBD.DevUse.InUse, nil
+}
+
+// getExistingBDWithFsUuid returns the blockdevice with matching FSUUID annotation from etcd
+func getExistingBDWithFsUuid(bd blockdevice.BlockDevice, bdAPIList *apis.BlockDeviceList) *apis.BlockDevice {
+	for _, bdAPI := range bdAPIList.Items {
+		if len(bd.FSInfo.FileSystemUUID) == 0 {
+			continue
+		}
+		fsUUID, ok := bdAPI.Annotations[internalFSUUIDAnnotation]
+		if !ok {
+			continue
+		}
+		if fsUUID == bd.FSInfo.FileSystemUUID {
+			return &bdAPI
+		}
+	}
+	return nil
+}
+
+// getExistingBDWithPartitionUUID returns the blockdevice with matching partition uuid annotation from etcd
+func getExistingBDWithPartitionUUID(bd blockdevice.BlockDevice, bdAPIList *apis.BlockDeviceList) *apis.BlockDevice {
+	for _, bdAPI := range bdAPIList.Items {
+		if len(bd.PartitionInfo.PartitionTableUUID) == 0 {
```

Comment:
We should return if bd.PartitionInfo.PartitionTableUUID is empty. Do we need to loop through bdAPIList in this case?

Reply:
Nope. will change it now.

Reply:
Done. @pawanpraka1 . Both partitionUUID check and FsUUID check has been moved out of the loop.